docs/man/plugins/sampler/hello_stream/hello_sampler.7.rst
.. _hello_sampler:

====================
hello_sampler
====================

:Date:   21 Aug 2021

NAME
====

hello_sampler - man page for the LDMS hello_sampler plugin

SYNOPSIS
========

| Within ldmsd_controller or a configuration file:
| config name=hello_sampler [ <attr>=<value> ]

DESCRIPTION
===========

With LDMS (Lightweight Distributed Metric Service), plugins for the
ldmsd (ldms daemon) are configured via ldmsd_controller or a
configuration file. The hello_sampler plugin does not actually sample,
but rather subscribes to an ldmsd_stream and writes the stream data to
the ldmsd logfile.

CONFIGURATION ATTRIBUTE SYNTAX
==============================

The hello_sampler plugin uses the sampler_base base class. This man page
covers only the configuration attributes, or those with default values,
specific to the this plugin; see ldms_sampler_base.man for the
attributes of the base class.

**config**
   | name=<plugin_name> stream=<stream>
   | configuration line

   name=<plugin_name>
      |
      | This MUST be hello_sampler.

   stream=<stream>
      |
      | Name of the stream to which to subscribe.

BUGS
====

No known bugs.

EXAMPLES
========

Within ldmsd_controller or a configuration file:

::

   load name=hello_sampler
   config name=hello_sampler producer=host1 instance=host1/hello_sampler stream=foo component_id=1
   start name=hello_sampler interval=1000000 offset=0

::

   > ./hello_publisher -x sock -h localhost -p 16000 -a munge -s foo -m "foo" -t str
   The data was successfully published.
   The server responded with 0

   > ./hello_publisher -x sock -h localhost -p 16000 -a munge -s foo -m "bar" -t str
   The data was successfully published.
   The server responded with 0


   In the log file of the ldmsd:
   > cat log.txt
   Mon May 04 19:44:05 2020: CRITICAL  : stream_type: STRING, msg: "foo", msg_len: 4, entity: (nil)
   Mon May 04 19:44:24 2020: CRITICAL  : stream_type: STRING, msg: "bar", msg_len: 4, entity: (nil)

   Note that the hello_streams sampler does not do a sample, instead it subscribes to the stream with a callback and prints out what it got off the stream.

SEE ALSO
========

:ref:`ldmsd(8) <ldmsd>`, :ref:`ldms_quickstart(7) <ldms_quickstart>`, :ref:`ldmsd_controller(8) <ldmsd_controller>`, :ref:`ldms_sampler_base(7) <ldms_sampler_base>`,
:ref:`ldmsd_stream_publish(7) <ldmsd_stream_publish>`, :ref:`stream_csv_store(7) <stream_csv_store>`


sos/NEWS
There's no news :D


ldms/src/sampler/hello_stream/hello_sampler.c
/* -*- c-basic-offset: 8 -*-
 * Copyright (c) 2019,2023 National Technology & Engineering Solutions
 * of Sandia, LLC (NTESS). Under the terms of Contract DE-NA0003525 with
 * NTESS, the U.S. Government retains certain rights in this software.
 * Copyright (c) 2019,2023 Open Grid Computing, Inc. All rights reserved.
 *
 * This software is available to you under a choice of one of two
 * licenses.  You may choose to be licensed under the terms of the GNU
 * General Public License (GPL) Version 2, available from the file
 * COPYING in the main directory of this source tree, or the BSD-type
 * license below:
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *      Redistributions of source code must retain the above copyright
 *      notice, this list of conditions and the following disclaimer.
 *
 *      Redistributions in binary form must reproduce the above
 *      copyright notice, this list of conditions and the following
 *      disclaimer in the documentation and/or other materials provided
 *      with the distribution.
 *
 *      Neither the name of Sandia nor the names of any contributors may
 *      be used to endorse or promote products derived from this software
 *      without specific prior written permission.
 *
 *      Neither the name of Open Grid Computing nor the names of any
 *      contributors may be used to endorse or promote products derived
 *      from this software without specific prior written permission.
 *
 *      Modified source versions must be plainly marked as such, and
 *      must not be misrepresented as being the original software.
 *
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
#define _GNU_SOURCE
#include <inttypes.h>
#include <unistd.h>
#include <sys/errno.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/inotify.h>
#include <time.h>
#include <pthread.h>
#include <strings.h>
#include <ctype.h>
#include <pwd.h>
#include <grp.h>
#include <ovis_json/ovis_json.h>
#include <assert.h>
#include <sched.h>
#include "ldms.h"
#include "ldmsd.h"

static ovis_log_t mylog;
static char *stream;

static const char *usage(struct ldmsd_plugin *self)
{
	return  "config name=hello_sampler producer=<producer_name> instance=<instance_name>\n"
		"         [stream=<stream_name>] [component_id=<component_id>] [perm=<permissions>]\n"
                "         [uid=<user_name>] [gid=<group_name>] [job_count=<job_length>]\n"
                "         [task_count=<task_length>]\n"
		"     producer      A unique name for the host providing the data\n"
		"     instance      A unique name for the metric set\n"
		"     stream        A stream name to subscribe the hello sampler to.\n"
		"                   Defaults to 'hello'\n";
}

static int sample(struct ldmsd_sampler *self)
{
	return 0;
}

static int hello_recv_cb(ldms_stream_event_t ev, void *arg)
{
	int rc = 0;
	const char *type = "UNKNOWN";
	if (ev->type != LDMS_STREAM_EVENT_RECV)
		return 0;
	switch (ev->recv.type) {
	case LDMS_STREAM_JSON:
		type = "JSON";
		break;
	case LDMS_STREAM_STRING:
		type = "STRING";
		break;
	}
	ovis_log(mylog, OVIS_LCRITICAL, "stream_type: %s, msg: \"%s\", "
					  "msg_len: %d, entity: %p\n",
	       type, ev->recv.data, ev->recv.data_len, ev->recv.json);
	return rc;
}

static int config(struct ldmsd_plugin *self, struct attr_value_list *kwl,
		  struct attr_value_list *avl)
{
	char *value;
	int rc = 0;

	value = av_value(avl, "stream");
	if (value)
		stream = strdup(value);
	else
		stream = strdup("hello_stream/hello");

	ldms_stream_subscribe(stream, 0, hello_recv_cb, self, "hello_sampler");

	return rc;
}

static void term(struct ldmsd_plugin *self)
{
}

static struct ldmsd_sampler hello_sampler = {
	.base = {
		.name = "hello_sampler",
		.type = LDMSD_PLUGIN_SAMPLER,
		.term = term,
		.config = config,
		.usage = usage,
	},
	.sample = sample
};

struct ldmsd_plugin *get_plugin()
{
	int rc;
	mylog = ovis_log_register("sampler.hello_sampler",
				"Message for the hello_sampler plugin");
	if (!mylog) {
		rc = errno;
		ovis_log(NULL, OVIS_LWARN, "Failed to create the log subsystem "
					"of 'hello_sampler' plugin. Error %d\n", rc);
	}
	return &hello_sampler.base;
}

static void __attribute__ ((constructor)) hello_sampler_init(void)
{
}

static void __attribute__ ((destructor)) hello_sampler_term(void)
{
}
